v3/src/components/data-display/components/legend/categorical-legend.tsx

```diff
@@ -1,351 +1,162 @@
 
-const coordinatesToCatIndex = (lod: Layout, numCategories: number, localPoint: { x: number, y: number }) => {
-    const {x, y} = localPoint,
-      col = Math.floor(x / lod.columnWidth),
-      row = Math.floor(y / (keySize + padding)),
-      catIndex = row * lod.numColumns + col
-    return catIndex >= 0 && catIndex < numCategories ? catIndex : -1
-  },
-  catLocation = (lod: Layout, catData: Key[], index: number) => {
-    return {
-      x: axisGap + catData[index].column * lod.columnWidth,
-      y: /*labelHeight + */catData[index].row * (keySize + padding)
-    }
-  }
+    const dataConfiguration = useDataConfigurationContext()
+    const dataDisplayLayout = useDataDisplayLayout()
+    const duration = useRef(0)
+    const keysElt = useRef(null)
 
-export const CategoricalLegend = observer(
-  function CategoricalLegend({layerIndex, setDesiredExtent}: IBaseLegendProps) {
-    const dataConfiguration = useDataConfigurationContext(),
-      tileWidth = useDataDisplayLayout().tileWidth,
-      categoriesRef = useRef<string[] | undefined>(),
-      categoryData = useRef<Key[]>([]),
-      layoutData = useRef<Layout>({
-          maxWidth: 0,
-          fullWidth: 0,
-          numColumns: 0,
-          numRows: 0,
-          columnWidth: 0
-        }
-      ),
-      dragInfo = useRef<DragInfo>({
-        indexOfCategory: -1,
-        initialOffset: {x: 0, y: 0},
-        currentDragPosition: {x: 0, y: 0}
-      }),
-      duration = useRef(0)
-    const prevCategoryIndex = useRef(0)
-    const
-      keysElt = useRef(null)
+    // useState guarantees the model will only be created once
+    // useMemo doesn't have that guarantee
+    const [legendModel] = useState(
+      () => new CategoricalLegendModel(dataConfiguration, dataDisplayLayout)
+    )
 
-    const setCategoryData = useCallback(() => {
-      if (categoriesRef.current) {
-        const newCategoryData = categoriesRef.current.map((cat: string, index) => {
-          return (
-          {
-            category: cat,
-            color: dataConfiguration?.getLegendColorForCategory(cat) || missingColor,
-            column: index % layoutData.current.numColumns,
-            index,
-            row: Math.floor(index / layoutData.current.numColumns)
-          })
-        })
-        categoryData.current = newCategoryData
-      }
-    }, [dataConfiguration])
+    // This is outside of the main autorun because it only needs to run when
+    // the number of categories changes or the max width of a category changes.
+    // Also the setDesiredExtent might cause extra re-renders, so it only
+    // runs when the desiredExtent actually changes
+    useEffect(function updateDesiredExtent() {
+      return mstReaction(
+        () => {
+          if (dataConfiguration?.placeCanHaveZeroExtent('legend')) {
+            return 0
+          }
+          const lod = legendModel.layoutData
+          return lod.numRows * (keySize + padding) + labelHeight + axisGap
```

Comment:
`lod` seems unnecessarily obtuse. Why not simply:
```suggestion
          return legendModel.layoutData.numRows * (keySize + padding) + labelHeight + axisGap
```